pennylane/math/ode.py

```diff
@@ -0,0 +1,246 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# TODO: This code is heavily inspired by https://github.com/google/jax/blob/main/jax/experimental/ode.py
+# need to make sure copyright is not infringed
+"""Fix step size ODE solver"""
+
+import warnings
+from functools import partial
+
+try:
+    import jax
+    import jax.numpy as jnp
+    from jax import linear_util as lu
+    from jax.flatten_util import ravel_pytree
+    from jax.experimental import host_callback
+except ImportError as e:
+    raise ImportError(
+        "Module jax is required for ``qml.math.odeint`` functionality. "
+        "You can install jax via: pip install jax"
+    ) from e
```

Comment:
Careful! Given that the `math` module is imported top-level, this error will be raised every time a user imports pennylane without `jax` installed.

There are 2 fixes for this:
1. Move the `try`/`except` inside the function were `jax` is used.
2. Instead of `raise ImportError`, create a `has_jax` variable that defaults to `True`, and set the variable to `False` if jax is not installed. Then you can check this variable inside the function that needs jax, and raise the error there.

Reply:
Thanks! Tested it in an environment without jax installed and works now as expected 👍